docs/proposals/20210705-bastion-vms.md

```diff
@@ -0,0 +1,209 @@
+    * [Test Plan](#test-plan)
+  * [Implementation History](#implementation-history)
+
+
+## Glossary
+
+* CAPZ - An abbreviation of Cluster API Provider Azure.
+* KISS - Keep It Simple, Stupid (see [Wikipedia](https://en.wikipedia.org/wiki/KISS_principle))
+
+## Summary
+
+Accessing virtual machines forming a Kubernetes cluster for administrative purposes is a potential security threat.
+A common pattern to provide SSH/RDP access to virtual machines without exposing them to the public internet is to use 
+a `Bastion host` or `Jump host`, a dedicated machine that is exposed to the internet acting as a bridge between the administrators'
+computers and the Kubernetes nodes (that can therefore stay behind a firewall).
+This CAEP aims at providing a way to bootstrap `Bastion Hosts` in a CAPZ cluster levereging existing Controllers as much as possible
+by reusing CRDs like `MachineDeployment` and `AzureMachineTemplate`.
+
+## Motivation
+
+By default CAPZ clusters expose one of the Master nodes to the internet making it possible to access all the other nodes in the cluster via SSH or RDP.
+
+As an alternative, it is possible to use the [`AzureBastion`](https://azure.microsoft.com/en-us/services/azure-bastion/) feature to SSH/RDP to nodes using a web interface through 
+the `Azure Portal` (see [documentation](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/book/src/topics/ssh-access.md)).
+
+Those solutions might not cover all use cases. As a matter of example:
+
+- Using one master node as a jump host might be not compliant to the security policies for some companies.
+- Using `AzureBastion` might be inpractical or unusable in some use cases (machine to machine access would be nearly impossible).
+
+### Goals
+
+- Allow users to deploy virtual machines as bastion hosts in multiple availability zones using a custom cloud init file.
+
+### Non-Goals/Future Work
+
+- Make it possible to disable the default behaviour of exposing one master node to the internet.
+
+## Proposal
+
+To adhere to the KISS principle the idea is to reuse existing reconciliation controllers for creating the `Bastion VM(s)`.
+In practice that means creating 3 resources:
+
+1. A `Secret` holding the `cloud-init` config file to bootstrap the bastion VM with the needed config (has to be provided by user and can't be defaulted in any way)
+2. An `AzureMachineTemplate` for each bastion that describes how the VM should be (size, OS, disks, etc). Can be defaulted.
+3. A `MachineDeployment` to bind the two aformentioned resources and make CAPI/CAPZ controllers reconcile and create the VMs. Can be defaulted.
+
+The suggestion is to add a new subfield to the `AzureCluster/spec/azureBastion` field like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: AzureCluster
+metadata:
+  name: cluster-name
+  namespace: default
+spec:
+  azureBastion:
+    bastionHosts:
+    - bootstrapDataSecretName: "<secret name>"
```

Comment:
this could be confused with the field `DataSecretName` same name in CABPK

Reply:
That's exactly what it is. That field is meant to be a reference to a secret that is identical to the secret that CABPK would create for a "traditional" Machine.